packages/@lwc/engine-dom/src/custom-elements/create-custom-element-vanilla.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isUndefined } from '@lwc/shared';
+import type { LifecycleCallback } from '@lwc/engine-core';
+
+const cachedConstructors = new Map<string, CustomElementConstructor>();
+const elementsUpgradedOutsideLWC = new WeakSet<HTMLElement>();
+let elementBeingUpgradedByLWC = false;
+
+// Creates a constructor that is intended to be used as a vanilla custom element, except that the upgradeCallback is
+// passed in to the constructor so LWC can reuse the same custom element constructor for multiple components.
+// Another benefit is that only LWC can create components that actually do anything – if you do
+// `customElements.define('x-foo')`, then you don't have access to the upgradeCallback, so it's a dummy custom element.
+// This class should be created once per tag name.
+const createUpgradableConstructor = (
+    connectedCallback: LifecycleCallback,
+    disconnectedCallback: LifecycleCallback
+) => {
+    // TODO [#2972]: this class should expose observedAttributes as necessary
+    return class UpgradableConstructor extends HTMLElement {
+        constructor(upgradeCallback: LifecycleCallback) {
+            super();
+            // If the element is not created using lwc.createElement(), e.g. `document.createElement('x-foo')`,
+            // then elementBeingUpgraded will be false
+            if (elementBeingUpgradedByLWC) {
+                upgradeCallback(this);
+            } else {
+                // keep track of elements that were not created by lwc.createElement,
+                // so we can ignore their lifecycle hooks
+                elementsUpgradedOutsideLWC.add(this);
+
+                if (!isUndefined(upgradeCallback)) {
+                    // This will only happen if someone is doing something zany like passing their own
+                    // custom upgradeCallback into the Ctor.
+                    throw new Error(
+                        'Failed to create custom element: upgradeCallback should either be undefined or registered'
+                    );
```

Comment:
This is maybe one of the oddest things I changed, but we weren't handling this case before, and this aligns with how the pivots registry works.

Basically it treats empty arguments in the constructor as a no-op (e.g. caused by `document.createElement('x-foo')`, we just create a dummy element and ignore). Whereas if you try to sneak in your own custom callback/constructor, then it throws an error because you're probably doing something nefarious.